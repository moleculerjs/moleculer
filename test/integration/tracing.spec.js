"use strict";

const _ = require("lodash");
const H = require("./helpers");

const POSTS = [
	{ id: 1, title: "Post 1", author: 3 },
	{ id: 2, title: "Post 2", author: 1 },
	{ id: 3, title: "Post 3", author: 2 }
	//{ id: 4, title: "Post 4", author: 3 },
	//{ id: 5, title: "Post 5", author: 4 },
];

const USERS = [
	{ id: 1, name: "Walter White" },
	{ id: 2, name: "Jesse Pinkman" },
	{ id: 3, name: "Hank Schrader" },
	{ id: 4, name: "Saul Goodman" },
	{ id: 5, name: "Mike Ehrmantraut" }
];

function getSpanFields(store) {
	return store.map(o => {
		o = _.omit(o, ["startTime", "startTicks", "duration", "finishTime"]);
		o.logs = o.logs.map(entry => _.omit(entry, ["elapsed", "time"]));
		return o;
	});
}

describe("Test Tracing feature with actions", () => {
	let idCounter = 1;

	const COMMON_SETTINGS = {
		logger: false,
		namespace: "tracing",
		//transporter: "NATS",
		tracing: {
			enabled: true,
			events: true,
			exporter: {
				type: "Event",
				options: {
					interval: 0
				}
			}
		},
		uidGenerator: broker => `${broker.nodeID}-${idCounter++}`
	};

	//const stacks = [];

	const STORE = [];

	const broker0 = H.createNode(_.defaultsDeep({ nodeID: "broker-0" }, COMMON_SETTINGS), [
		{
			name: "tracing-collector",
			events: {
				"$tracing.spans": {
					tracing: false,
					handler(ctx) {
						STORE.push(...ctx.params);
					}
				}
				/*"**": {
					tracing: false,
					handler(ctx) {
						stacks.push(`${ctx.broker.nodeID}:${ctx.eventName}`);
					}
				}*/
			}
		}
	]);

	const broker1 = H.createNode(_.defaultsDeep({ nodeID: "broker-1" }, COMMON_SETTINGS), [
		{
			name: "posts",
			actions: {
				list: {
					async handler(ctx) {
						const posts = _.cloneDeep(POSTS);

						await Promise.all(
							posts.map(async post => {
								const author = await ctx.call("users.get", { id: post.author });
								post.author = author; //eslint-disable-line
								return post;
							})
						);

						return posts;
					}
				}
			},
			events: {
				async "comments.removed"(ctx) {
					const span1 = ctx.startSpan("update posts");
					ctx.broadcast("post.updated");
					ctx.finishSpan(span1);

					await broker1.Promise.delay(500);

					const span2 = ctx.startSpan("update others");
					ctx.broadcast("user.updated");
					ctx.finishSpan(span2);
				},

				"post.updated"(ctx) {
					ctx.span.log("some changes", { a: 5 });
				}
			}
		}
	]);

	const broker2 = H.createNode(_.defaultsDeep({ nodeID: "broker-2" }, COMMON_SETTINGS), [
		{
			name: "users",
			actions: {
				get: {
					tracing: {
						spanName: ctx => `Get user by ID: ${ctx.params.id}`,
						tags: {
							response: ["name"]
						}
					},

					async handler(ctx) {
						let user = USERS.find(user => user.id == ctx.params.id);
						if (user) {
							const span = ctx.startSpan("cloning", {
								tags: {
									userID: user.id
								}
							});
							user = _.cloneDeep(user);
							Promise.delay(5);
							ctx.finishSpan(span);

							user.friends = await ctx.call("friends.count", { id: user.id });
						}
						return user;
					}
				}
			},
			events: {
				"user.updated": {
					tracing: {
						spanName: "User updated event"
					},
					async handler(ctx) {
						const span1 = ctx.startSpan("updating user");
						// TODO: not perfect. Its parent is the event span and not span1
						await ctx.call("friends.count", { userID: 2 });
						ctx.finishSpan(span1);
					}
				}
			}
		}
	]);

	const broker3 = H.createNode(_.defaultsDeep({ nodeID: "broker-3" }, COMMON_SETTINGS), [
		{
			name: "friends",
			actions: {
				count: {
					tracing: {
						tags: {
							meta: ["user.name"]
						}
					},
					async handler(ctx) {
						return ctx.params.id * 2;
					}
				}
			}
		}
	]);

	beforeAll(() =>
		Promise.all([
			broker0.start(),
			broker1.start(),
			broker2.start(),
			broker3.start(),
			Promise.resolve().delay(2000)
		])
	);

	afterAll(() => Promise.all([broker0.stop(), broker1.stop(), broker2.stop(), broker3.stop()]));

	it("should generate action spans", async () => {
		idCounter = 1;

		const res = await broker1.call("posts.list", null, {
			meta: {
				user: {
					id: 100,
					name: "Superuser"
				}
			}
		});

		expect(res).toMatchSnapshot();

		//Promise.delay(200);
		STORE.sort((a, b) => a.startTicks - b.startTicks);

		const spans = getSpanFields(STORE);

		expect(spans).toMatchSnapshot();
	});

	it("should generate event spans", async () => {
		idCounter = 1;
		STORE.length = 0;
		//stacks.length = 0;

		await broker1.emit("comments.removed", null, {
			meta: {
				user: {
					id: 100,
					name: "Superuser"
				}
			}
		});

		await Promise.delay(500);

		STORE.sort((a, b) => a.startTicks - b.startTicks);

		const spans = getSpanFields(STORE);

		//console.log(stacks);

		expect(spans).toMatchSnapshot();
	});
});
