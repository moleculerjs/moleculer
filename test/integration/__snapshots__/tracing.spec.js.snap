// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Tracing feature with actions should generate action spans 1`] = `
Array [
  Object {
    "author": Object {
      "friends": 6,
      "id": 3,
      "name": "Hank Schrader",
    },
    "id": 1,
    "title": "Post 1",
  },
  Object {
    "author": Object {
      "friends": 2,
      "id": 1,
      "name": "Walter White",
    },
    "id": 2,
    "title": "Post 2",
  },
  Object {
    "author": Object {
      "friends": 4,
      "id": 2,
      "name": "Jesse Pinkman",
    },
    "id": 3,
    "title": "Post 3",
  },
]
`;

exports[`Test Tracing feature with actions should generate action spans 2`] = `
Array [
  Object {
    "error": null,
    "id": "broker-1-1",
    "logs": Array [],
    "name": "action 'posts.list'",
    "parentID": null,
    "priority": 5,
    "sampled": true,
    "service": Object {
      "fullName": "posts",
      "name": "posts",
    },
    "tags": Object {
      "action": Object {
        "name": "posts.list",
        "rawName": "list",
      },
      "callerNodeID": "broker-1",
      "callingLevel": 1,
      "fromCache": false,
      "nodeID": "broker-1",
      "options": Object {},
      "params": null,
      "remoteCall": false,
      "requestID": "broker-1-1",
    },
    "traceID": "broker-1-1",
    "type": "action",
  },
  Object {
    "error": null,
    "id": "broker-1-2",
    "logs": Array [],
    "name": "Get user by ID: 3",
    "parentID": "broker-1-1",
    "priority": 5,
    "sampled": true,
    "service": Object {
      "fullName": "users",
      "name": "users",
    },
    "tags": Object {
      "action": Object {
        "name": "users.get",
        "rawName": "get",
      },
      "callerNodeID": "broker-1",
      "callingLevel": 2,
      "fromCache": false,
      "nodeID": "broker-2",
      "options": Object {
        "retries": null,
        "timeout": 0,
      },
      "params": Object {
        "id": 3,
      },
      "remoteCall": true,
      "requestID": "broker-1-1",
      "response": Object {
        "name": "Hank Schrader",
      },
    },
    "traceID": "broker-1-1",
    "type": "action",
  },
  Object {
    "error": null,
    "id": "broker-2-4",
    "logs": Array [],
    "name": "cloning",
    "parentID": "broker-1-2",
    "priority": 5,
    "sampled": true,
    "service": Object {
      "fullName": "users",
      "name": "users",
    },
    "tags": Object {
      "userID": 3,
    },
    "traceID": "broker-1-1",
    "type": "custom",
  },
  Object {
    "error": null,
    "id": "broker-2-9",
    "logs": Array [],
    "name": "action 'friends.count'",
    "parentID": "broker-1-2",
    "priority": 5,
    "sampled": true,
    "service": Object {
      "fullName": "friends",
      "name": "friends",
    },
    "tags": Object {
      "action": Object {
        "name": "friends.count",
        "rawName": "count",
      },
      "callerNodeID": "broker-2",
      "callingLevel": 3,
      "fromCache": false,
      "meta": Object {
        "user": Object {
          "name": "Superuser",
        },
      },
      "nodeID": "broker-3",
      "options": Object {
        "retries": null,
        "timeout": 0,
      },
      "params": Object {
        "id": 3,
      },
      "remoteCall": true,
      "requestID": "broker-1-1",
    },
    "traceID": "broker-1-1",
    "type": "action",
  },
  Object {
    "error": null,
    "id": "broker-1-11",
    "logs": Array [],
    "name": "Get user by ID: 1",
    "parentID": "broker-1-1",
    "priority": 5,
    "sampled": true,
    "service": Object {
      "fullName": "users",
      "name": "users",
    },
    "tags": Object {
      "action": Object {
        "name": "users.get",
        "rawName": "get",
      },
      "callerNodeID": "broker-1",
      "callingLevel": 2,
      "fromCache": false,
      "nodeID": "broker-2",
      "options": Object {
        "retries": null,
        "timeout": 0,
      },
      "params": Object {
        "id": 1,
      },
      "remoteCall": true,
      "requestID": "broker-1-1",
      "response": Object {
        "name": "Walter White",
      },
    },
    "traceID": "broker-1-1",
    "type": "action",
  },
  Object {
    "error": null,
    "id": "broker-2-13",
    "logs": Array [],
    "name": "cloning",
    "parentID": "broker-1-11",
    "priority": 5,
    "sampled": true,
    "service": Object {
      "fullName": "users",
      "name": "users",
    },
    "tags": Object {
      "userID": 1,
    },
    "traceID": "broker-1-1",
    "type": "custom",
  },
  Object {
    "error": null,
    "id": "broker-2-18",
    "logs": Array [],
    "name": "action 'friends.count'",
    "parentID": "broker-1-11",
    "priority": 5,
    "sampled": true,
    "service": Object {
      "fullName": "friends",
      "name": "friends",
    },
    "tags": Object {
      "action": Object {
        "name": "friends.count",
        "rawName": "count",
      },
      "callerNodeID": "broker-2",
      "callingLevel": 3,
      "fromCache": false,
      "meta": Object {
        "user": Object {
          "name": "Superuser",
        },
      },
      "nodeID": "broker-3",
      "options": Object {
        "retries": null,
        "timeout": 0,
      },
      "params": Object {
        "id": 1,
      },
      "remoteCall": true,
      "requestID": "broker-1-1",
    },
    "traceID": "broker-1-1",
    "type": "action",
  },
  Object {
    "error": null,
    "id": "broker-1-20",
    "logs": Array [],
    "name": "Get user by ID: 2",
    "parentID": "broker-1-1",
    "priority": 5,
    "sampled": true,
    "service": Object {
      "fullName": "users",
      "name": "users",
    },
    "tags": Object {
      "action": Object {
        "name": "users.get",
        "rawName": "get",
      },
      "callerNodeID": "broker-1",
      "callingLevel": 2,
      "fromCache": false,
      "nodeID": "broker-2",
      "options": Object {
        "retries": null,
        "timeout": 0,
      },
      "params": Object {
        "id": 2,
      },
      "remoteCall": true,
      "requestID": "broker-1-1",
      "response": Object {
        "name": "Jesse Pinkman",
      },
    },
    "traceID": "broker-1-1",
    "type": "action",
  },
  Object {
    "error": null,
    "id": "broker-2-22",
    "logs": Array [],
    "name": "cloning",
    "parentID": "broker-1-20",
    "priority": 5,
    "sampled": true,
    "service": Object {
      "fullName": "users",
      "name": "users",
    },
    "tags": Object {
      "userID": 2,
    },
    "traceID": "broker-1-1",
    "type": "custom",
  },
  Object {
    "error": null,
    "id": "broker-2-27",
    "logs": Array [],
    "name": "action 'friends.count'",
    "parentID": "broker-1-20",
    "priority": 5,
    "sampled": true,
    "service": Object {
      "fullName": "friends",
      "name": "friends",
    },
    "tags": Object {
      "action": Object {
        "name": "friends.count",
        "rawName": "count",
      },
      "callerNodeID": "broker-2",
      "callingLevel": 3,
      "fromCache": false,
      "meta": Object {
        "user": Object {
          "name": "Superuser",
        },
      },
      "nodeID": "broker-3",
      "options": Object {
        "retries": null,
        "timeout": 0,
      },
      "params": Object {
        "id": 2,
      },
      "remoteCall": true,
      "requestID": "broker-1-1",
    },
    "traceID": "broker-1-1",
    "type": "action",
  },
]
`;

exports[`Test Tracing feature with actions should generate event spans 1`] = `
Array [
  Object {
    "error": null,
    "id": "broker-1-2",
    "logs": Array [],
    "name": "event 'comments.removed' in 'posts'",
    "parentID": null,
    "priority": 5,
    "sampled": true,
    "service": Object {
      "fullName": "posts",
      "name": "posts",
    },
    "tags": Object {
      "callerNodeID": "broker-1",
      "callingLevel": 1,
      "event": Object {
        "name": "comments.removed",
      },
      "eventName": "comments.removed",
      "eventType": "emit",
      "nodeID": "broker-1",
      "params": null,
      "remoteCall": false,
      "requestID": "broker-1-1",
    },
    "traceID": "broker-1-1",
    "type": "event",
  },
  Object {
    "error": null,
    "id": "broker-1-3",
    "logs": Array [],
    "name": "update posts",
    "parentID": "broker-1-2",
    "priority": 5,
    "sampled": true,
    "service": Object {
      "fullName": "posts",
      "name": "posts",
    },
    "tags": Object {},
    "traceID": "broker-1-1",
    "type": "custom",
  },
  Object {
    "error": null,
    "id": "broker-1-6",
    "logs": Array [
      Object {
        "fields": Object {
          "a": 5,
        },
        "name": "some changes",
      },
    ],
    "name": "event 'post.updated' in 'posts'",
    "parentID": "broker-1-3",
    "priority": 5,
    "sampled": true,
    "service": Object {
      "fullName": "posts",
      "name": "posts",
    },
    "tags": Object {
      "callerNodeID": "broker-1",
      "callingLevel": 2,
      "event": Object {
        "name": "post.updated",
      },
      "eventName": "post.updated",
      "eventType": "broadcastLocal",
      "nodeID": "broker-1",
      "params": null,
      "remoteCall": false,
      "requestID": "broker-1-1",
    },
    "traceID": "broker-1-1",
    "type": "event",
  },
  Object {
    "error": null,
    "id": "broker-1-15",
    "logs": Array [],
    "name": "update others",
    "parentID": "broker-1-2",
    "priority": 5,
    "sampled": true,
    "service": Object {
      "fullName": "posts",
      "name": "posts",
    },
    "tags": Object {},
    "traceID": "broker-1-1",
    "type": "custom",
  },
  Object {
    "error": null,
    "id": "broker-2-19",
    "logs": Array [],
    "name": "User updated event",
    "parentID": "broker-1-15",
    "priority": 5,
    "sampled": true,
    "service": Object {
      "fullName": "users",
      "name": "users",
    },
    "tags": Object {
      "callerNodeID": "broker-1",
      "callingLevel": 2,
      "event": Object {
        "name": "user.updated",
      },
      "eventName": "user.updated",
      "eventType": "broadcast",
      "nodeID": "broker-2",
      "params": null,
      "remoteCall": true,
      "requestID": "broker-1-1",
    },
    "traceID": "broker-1-1",
    "type": "event",
  },
  Object {
    "error": null,
    "id": "broker-2-20",
    "logs": Array [],
    "name": "updating user",
    "parentID": "broker-2-19",
    "priority": 5,
    "sampled": true,
    "service": Object {
      "fullName": "users",
      "name": "users",
    },
    "tags": Object {},
    "traceID": "broker-1-1",
    "type": "custom",
  },
  Object {
    "error": null,
    "id": "broker-2-21",
    "logs": Array [],
    "name": "action 'friends.count'",
    "parentID": "broker-2-20",
    "priority": 5,
    "sampled": true,
    "service": Object {
      "fullName": "friends",
      "name": "friends",
    },
    "tags": Object {
      "action": Object {
        "name": "friends.count",
        "rawName": "count",
      },
      "callerNodeID": "broker-2",
      "callingLevel": 3,
      "fromCache": false,
      "meta": Object {
        "user": Object {
          "name": "Superuser",
        },
      },
      "nodeID": "broker-3",
      "options": Object {
        "retries": null,
        "timeout": 0,
      },
      "params": Object {
        "userID": 2,
      },
      "remoteCall": true,
      "requestID": "broker-1-1",
    },
    "traceID": "broker-1-1",
    "type": "action",
  },
]
`;
